plugins/scaffolder-backend/src/lib/templating/SecureTemplater.ts

```diff
@@ -0,0 +1,141 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { VM } from 'vm2';
+import { resolvePackagePath } from '@backstage/backend-common';
+import fs from 'fs-extra';
+import { RepoSpec } from '../../scaffolder/actions/builtin/publish/util';
+
+const mkScript = (nunjucksSource: string) => `
+const { render, renderCompat } = (() => {
+  const module = {};
+  const process = { env: {} };
+  const require = (pkg) => { if (pkg === 'events') { return function (){}; }};
+
+  ${nunjucksSource}
+
+  const env = module.exports.configure({
+    autoescape: false,
+    tags: {
+      variableStart: '\${{',
+      variableEnd: '}}',
+    },
+  });
+
+  const compatEnv = module.exports.configure({
+    autoescape: false,
+    tags: {
+      variableStart: '{{',
+      variableEnd: '}}',
+    },
+  });
+  compatEnv.addFilter('jsonify', compatEnv.getFilter('dump'));
+
+  if (typeof parseRepoUrl !== 'undefined') {
+    const safeHelperRef = parseRepoUrl;
+
+    env.addFilter('parseRepoUrl', repoUrl => {
+      return JSON.parse(safeHelperRef(repoUrl))
+    });
+    env.addFilter('projectSlug', repoUrl => {
+      const { owner, repo } = JSON.parse(safeHelperRef(repoUrl));
+      return owner + '/' + repo;
+    });
+  }
+
+  let uninstallCompat = undefined;
+
+  function render(str, values) {
+    try {
+      if (uninstallCompat) {
+        uninstallCompat();
+        uninstallCompat = undefined;
+      }
+      return env.renderString(str, JSON.parse(values));
+    } catch (error) {
+      // Make sure errors don't leak anything
+      throw new Error(String(error.message));
+    }
+  }
+
+  function renderCompat(str, values) {
+    try {
+      if (!uninstallCompat) {
+        uninstallCompat = module.exports.installJinjaCompat();
+      }
+      return compatEnv.renderString(str, JSON.parse(values));
+    } catch (error) {
+      // Make sure errors don't leak anything
+      throw new Error(String(error.message));
+    }
+  }
+
+  return { render, renderCompat };
+})();
+`;
+
+export interface SecureTemplaterOptions {
+  /* Optional implementation of the parseRepoUrl filter */
+  parseRepoUrl?(repoUrl: string): RepoSpec;
+
+  /* Enables jinja compatibility and the "jsonify" filter */
+  cookiecutterCompat?: boolean;
+}
+
+export type SecureTemplateRenderer = (
+  template: string,
+  values: unknown,
+) => string;
+
+export class SecureTemplater {
+  static async loadRenderer(options: SecureTemplaterOptions = {}) {
+    const { parseRepoUrl, cookiecutterCompat } = options;
+    let sandbox = undefined;
+
+    if (parseRepoUrl) {
+      sandbox = {
+        parseRepoUrl: (url: string) => JSON.stringify(parseRepoUrl(url)),
+      };
+    }
+
+    const vm = new VM({ timeout: 1000, sandbox });
```

Comment:
wondering about this timeout, i hope that does not apply to the total template execution time?

Reply:
We are not doing anything async. Removed timeout.